frame/base/bli_riscv_detect_arch.h

```diff
@@ -0,0 +1,148 @@
+   met:
+    - Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    - Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in the
+      documentation and/or other materials provided with the distribution.
+    - Neither the name(s) of the copyright holder(s) nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+   "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+   LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+   A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+   HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+   DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+   THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+   (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+   OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+
+*/
+
+/* Construct a RISC-V architecture string based on available features. */
+
+#if __riscv
+
+#if __riscv_arch_test
+
+#if __riscv_i
+#define RISCV_I i
+#else
+#define RISCV_I
+#endif
+
+#if __riscv_e
+#define RISCV_E e
+#else
+#define RISCV_E
+#endif
+
+#if __riscv_m
+#define RISCV_M m
+#else
+#define RISCV_M
+#endif
+
+#if __riscv_a
+#define RISCV_A a
+#else
+#define RISCV_A
+#endif
+
+#if __riscv_f
+#define RISCV_F f
+#else
+#define RISCV_F
+#endif
+
+#if __riscv_d
+#define RISCV_D d
+#else
+#define RISCV_D
+#endif
+
+#if __riscv_c
+#define RISCV_C c
+#else
+#define RISCV_C
+#endif
+
+#if __riscv_p
+#define RISCV_P p
+#else
+#define RISCV_P
+#endif
+
+#if __riscv_v
+#define RISCV_V v
+#else
+#define RISCV_V
+#endif
+
+#else /* __riscv_arch_test */
+
+/* We assume I and E are exclusive when __riscv_arch_test isn't defined */
+#if __riscv_32e
+#define RISCV_I
+#define RISCV_E e
+#else
+#define RISCV_I i
+#define RISCV_E
+#endif
+
+#if __riscv_flen >= 32
+#define RISCV_F f
+#else
+#define RISCV_F
+#endif
+
+#if __riscv_flen >= 64
+#define RISCV_D d
+#else
+#define RISCV_D
+#endif
+
+/* Cannot test for M, A, C, P if __riscv_arch_test isn't defined */
+#define RISCV_M
+#define RISCV_A
+#define RISCV_C
+#define RISCV_P
+
+#if __riscv_vector
+#define RISCV_V v
+#else
+#define RISCV_V
+#endif
+
+#endif /* __riscv_arch_test */
+
+#define CAT2(a,b) a##b
+#define CAT(a,b) CAT2(a,b)
+
+#if __riscv_arch_test && ( __riscv_i && __riscv_m && __riscv_a && \
+						   __riscv_f && __riscv_d && !__riscv_e)
+
+CAT(rv, CAT(__riscv_xlen, CAT(g, CAT(RISCV_C, CAT(RISCV_P, RISCV_V)))))
```

Comment:
I am not sure if this is fully correct. My understanding is that `g` subsumes all of the extensions, but covers more.

https://five-embeddev.com/riscv-isa-manual/latest/naming.html#instruction-set-extension-names:
> We have also defined an abbreviation “G” to represent the “IMAFDZicsr_Zifencei” base and extensions, as this is intended to represent our standard general-purpose ISA.

Reply:
I don't know how to test for `Zicsr` or `Zfencei` (or the all-encompassing G pseudo-extension) with the C API. I can remove the G shorthand and just list the individual extensions which we can detect.